stacks-signer/src/stacks_client.rs

```diff
@@ -158,143 +193,191 @@ impl StacksClient {
+    /// Helper function to retrieve the next possible nonce for the signer from the stacks node
+    #[allow(dead_code)]
+    fn get_next_possible_nonce(&self) -> Result<u64, ClientError> {
+        //FIXME: use updated RPC call to get mempool nonces. Depends on https://github.com/stacks-network/stacks-blockchain/issues/4000
+        todo!("Get the next possible nonce from the stacks node");
+    }
 
-        let mut tx_signer = StacksTransactionSigner::new(&unsigned_tx);
-        tx_signer
-            .sign_origin(&self.stacks_private_key)
-            .map_err(|_| ClientError::SignatureGenerationFailure)?;
-        if let (Some(payer), Some(_)) = (payer, payer_nonce) {
-            tx_signer
-                .sign_sponsor(payer)
-                .map_err(|_| ClientError::SponsorSignatureGenerationFailure)?;
+    /// Helper function to retrieve the pox contract address and name from the stacks node
+    fn get_pox_contract(&self) -> Result<(StacksAddress, ContractName), ClientError> {
+        // Check if we have overwritten the pox contract ID in the config
+        if let Some(pox_contract) = self.pox_contract_id.clone() {
+            return Ok((pox_contract.issuer.into(), pox_contract.name));
         }
-
-        let Some(tx) = tx_signer.get_tx() else {
-            return Err(ClientError::SignatureGenerationFailure);
+        // TODO: we may want to cache the pox contract inside the client itself (calling this function once on init)
+        // https://github.com/stacks-network/stacks-blockchain/issues/4005
+        let send_request = || {
+            self.stacks_node_client
+                .get(self.pox_path())
+                .send()
+                .map_err(backoff::Error::transient)
         };
+        let response = retry_with_exponential_backoff(send_request)?;
+        if !response.status().is_success() {
+            return Err(ClientError::RequestFailure(response.status()));
+        }
+        let json_response = response.json::<serde_json::Value>()?;
+        let entry = "contract_id";
+        let contract_id_string = json_response
+            .get(entry)
+            .and_then(|id: &serde_json::Value| id.as_str())
+            .ok_or_else(|| ClientError::InvalidJsonEntry(entry.to_string()))?;
+        let id = QualifiedContractIdentifier::parse(contract_id_string).unwrap();
+        Ok((id.issuer.into(), id.name))
+    }
+
+    /// Helper function that attempts to deserialize a clarity hex string as the aggregate public key
+    fn parse_aggregate_public_key(&self, hex: &str) -> Result<Option<Point>, ClientError> {
+        let public_key_clarity_value = ClarityValue::try_deserialize_hex_untyped(hex)?;
+        if let ClarityValue::Optional(optional_data) = public_key_clarity_value.clone() {
+            if let Some(ClarityValue::Sequence(SequenceData::Buffer(public_key))) =
+                optional_data.data.map(|boxed| *boxed)
+            {
+                let mut bytes = [0_u8; 32];
```

Comment:
You're going to want to check the length of `public_key`.  It might not be 32 bytes; a `(buff 32)` is _at most_ 32 bytes.

Reply:
Oh good catch! Fixed :D